### PR TITLE
Reserve Bounties

### LINKED ISSUE #2037: BUILD - Reserve Bounties
### User Story

[comment]: # (As a <user type>, I want to <task> so that <goal>.)
As a funder, I want to fund and send a bounty directly to a certain group of individuals (or perhaps, just one individual).

### Why Is this Needed
This is needed so that select group of individuals can work on a project.

[comment]: # (Describe the problem and why this task is needed. Provide description of the current state, what you would like to happen, and what actually happen)
*Summary*: Right now, I sometimes get unnecessary noise on bounties that I would rather send directly to a smaller group of people that I know from prior work on the platform. 

### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Feature

### Current Behavior
[comment]: # (Describe what actually happened.)
Post a bounty on the main issue explorer, make it approval-only, and approve the worker of my choosing. If I don't get to approve within 24 hours, perhaps deal with auto-approvals of people I didn't want to get involved. 

### Definition of Done
[comment]: # (Any other information that would be useful, bullets are helpful.)

**Funder Form**
- [ ] On the funder form under Permissions Field add the following: Reserved for ___________________.
- [ ] User should be able to type in a username in the input field

**Issue Details**
- [ ] The Issue Details should display that this issue is Reserved for X Contributor with the contributor's icon, linking to their profile.
- [ ] Only the contributor(s) that have been selected to work this issue can click start work.
- [ ] Group information and line the information up like this (Will Tip Extra)
screen shot 2018-08-27 at 7 43 16 am

**Email**
- [ ] User(s) should be notified that an issue has been reserved for them to work on.

### Reviewed by 
[comment]: # (Any other information that would be useful, content, screenshots, etc.)

@owocki @PixelantDesign @thelostone-mc Thoughts?

### PR DESCRIPTION
**Pull request by mul1sh**
##### Description

This PR allows funders to reserve bounties for particular users for 24 hours.

**Funder Form**
[x] On the funder form under Permissions Field add the following: Reserved for _.
[x] User should be able to type in a username in the input field

**Issue Details**
[x] The Issue Details should display that this issue is Reserved for X Contributor with the contributor's icon, linking to their profile.
[x] Only the contributor(s) that have been selected to work this issue can click start work.
[x] Group information and line the information up like this (Will Tip Extra)

**Email**
[x] User(s) should be notified that an issue has been reserved for them to work on.

##### Checklist

- [x] linter status: 100% pass
- [x] changes don't break existing behavior
- [x] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

##### Affected core subsystem(s)

Web

##### Testing

Yes I have tested this PR, and it will not break anything when merged into master 🙂 

##### Refers/Fixes
Fixes #2037

### CODE REVIEW

File: app/assets/v2/css/bounty.css [hunk 1]
```diff
@@ -160,14 +164,40 @@ body {
...
 
+
 .bounty-info-heading {
   color: #0D0764;
 }
 
+.bounty-info-row-div {
+  width: 100%;
+}
+.bounty-info-row-span-left{
```
Comment (thelostone-mc) [inline app/assets/v2/css/bounty.css]: could we add a new line here?

File: app/assets/v2/css/bounty.css [hunk 2]
```diff
@@ -160,14 +164,40 @@ body {
...
+.bounty-info-row-span-right{
+  float: right;
+}
+
+#bounty_reserved_for {
+  float: right;
+  margin-right: 70px;
+}
+
+
```
Comment (thelostone-mc) [inline app/assets/v2/css/bounty.css]: do away with the extra line :P

**Comment by thelostone-mc** [comment]
@mul1sh could you throw in a screen recording / screenshots of the UI changes so that's easier for @PixelantDesign to review

File: app/assets/v2/js/pages/bounty_details.js [hunk 3]
```diff
@@ -682,8 +708,35 @@ var show_extend_deadline_modal = function() {
...
+      $('#bounty_reserved_for').html(reservedForHtmlLink + reservedForAvatar);
+    } else {
+      $('#bounty_reserved_for').css('display', 'none');
+      $('#bounty_reserved_for_label').css('display', 'none');
+    }
+  } else {
+    $('#bounty_reserved_for').css('display', 'none');
+    $('#bounty_reserved_for_label').css('display', 'none');
+  }
+};
```
Comment (thelostone-mc) [inline app/assets/v2/js/pages/bounty_details.js]: Can we use `let` , `const` as opposed to `var`? 
We are starting on this

File: app/assets/v2/js/pages/bounty_details.js [hunk 4]
```diff
@@ -529,7 +532,30 @@ var attach_work_actions = function() {
   $('body').delegate('a[href="/interested"], a[href="/uninterested"], a[href="/extend-deadlines"]', 'click', function(e) {
     e.preventDefault();
     if ($(this).attr('href') == '/interested') {
-      show_interest_modal.call(this);
+      if (currentLoggedInUser.length > 0 && reservedForUsername.length > 0) {
```
Comment (thelostone-mc) [inline app/assets/v2/js/pages/bounty_details.js]: This ensures if a user clicks on start work (which is reserved for someone else ), he'd see a modal with the waiting message ( As an end user I would be frustrated to click a bounty and to only know it's reserved for someone else )

Should we hide the start work button for other contributors (aka do a check on the `details.html`) until the issue is open to everyone? 

cc @PixelantDesign @mbeacom @SaptakS

File: app/assets/v2/css/bounty.css [hunk 5]
```diff
@@ -160,14 +164,40 @@ body {
...
 .bounty-info-text {
   color: #717171;
 }
 
+.col-4.bounty-info{
+  margin-left: 20px;
+}
+
+.col-4.bounty-info span.bounty-info-text.bounty-info-row-span-right {
+  margin-right: 70px;
```
Comment (thelostone-mc) [inline app/assets/v2/css/bounty.css]: could we use `em` as opposed to `px` ?

File: app/assets/v2/css/bounty.css [hunk 6]
```diff
@@ -160,14 +164,40 @@ body {
...
+  float: right;
+  margin-right: 70px;
+}
+
+
 .bounty-info-text {
   color: #717171;
 }
 
+.col-4.bounty-info{
```
Comment (thelostone-mc) [inline app/assets/v2/css/bounty.css]: `-info {` spacing needed

File: app/assets/v2/js/pages/bounty_details.js [hunk 7]
```diff
@@ -159,6 +160,8 @@ var callbacks = {
   'bounty_owner_email': email_ize,
   'experience_level': unknown_if_empty,
   'project_length': unknown_if_empty,
+  'bounty_reserved_for': unknown_if_empty,
+  'bounty_reserved_for_creation_date': unknown_if_empty,
```
Comment (thelostone-mc) [inline app/assets/v2/js/pages/bounty_details.js]: what is this being used for ?

**Comment by PixelantDesign** [comment]
@thelostone-mc 
What would be the reasoning to not enable for contest and cooperative? Those workers need to get paid as well, no?

**Comment by vs77bb** [comment]
@PixelantDesign I think @thelostone-mc was thinking bounties that are contests / cooperative by definition *can't* be reserved for one person. 

I think that makes sense -- it should only be for traditional bounties.

**Comment by mul1sh** [comment]
@thelostone-mc here is a link to a video showing the UI changes https://vimeo.com/289170255 . I have implemented the changes requested and so I hope its shows everything clearly. I've also pushed some bug fixes as requested so kindly check on them as well 😃 thanks.

**Comment by mul1sh** [comment]
@PixelantDesign not really, let me reduce the space

**Comment by PixelantDesign** [comment]
looks good thanks!

**Comment by mul1sh** [comment]
@PixelantDesign sorry for that, the above fix had some issues but i've pushed the correct one

**Comment by mul1sh** [comment]
@thelostone-mc  any more changes to be made in this PR? thanks 🙂

**Comment by vs77bb** [comment]
@mul1sh @thelostone-mc My only other comment: can we automatically put this issue in 'Work Started' and list the 'Reserved' individual as under 'Work Started'? 

This will avoid confusion on the Issue Explorer as the issue won't show up as 'Open'.

Outside of the scope for this ticket (it wasn't listed in #2037), but perhaps in a v2 we can send an e-mail / inbox notification to the Gitcoiner in question to have them **confirm** they are willing to work the bounty. For v1, I think we can assume that the Reserve bounty will be used in cases they know that a particular individual will do the work.

**Comment by mul1sh** [comment]
@thelostone-mc thanks for the feedback, sorting it now

**Comment by mul1sh** [comment]
@vs77bb sure, i'll re-implement it this way

**Comment by PixelantDesign** [comment]
@mul1sh is there a reason this was closed?

**Comment by mul1sh** [comment]
@PixelantDesign apologies for closing this, but I explained why [here](https://github.com/gitcoinco/web/pull/2275#issuecomment-424392557)  and I also re-opened the PR   [here](https://github.com/gitcoinco/web/pull/2275) which is basically a duplicate of this PR.

**Comment by PixelantDesign** [comment]
ok! thanks! I'll close

**Pull request closed without merging**